### PR TITLE
Rewrite - a new approach

### PR DESCRIPTION
**Pull request by rboulton**
This is an attempt to address several problems with the existing mix of shell and ruby scripts. Firstly, the mix of shell and ruby is a bit hard to follow.

Secondly, the log files aren't actually for a single day, and sometimes we have several days logs in a single file (eg, because logrotate breaks for a bit), so we need to parse the timestamps in the file - which includes being able to parse timezone information, and then merge the resulting counts from different files to get the correct counts for a single day.

Thirdly, we want to be able to ensure that we have accurate counts for the log files which exist, even if they get modified (or changed to a compressed form) - so this design moves to having a single script for counting all the log files which are present, which will skip log files which have already been counted: it ensures that a log file which is later modified is re-counted by tracking the file size when the log was counted.  This script will be run nightly to update the derived counts.

Finally, and perhaps most importantly, we may not be able to keep the GOV.UK logs indefinitely, due to data size and potential privacy implications, but a digest of just the counts of accesses to URLs in a time period will take up much less space, have fewer privacy risks, and be almost as useful for data analysis. We expect in future to want to do more than just look for successful requests in the logs, so this commit takes the approach of first running a pure count of the combination of hour of day, HTTP method, path, and resulting status code.

Later stages of processing are then done to using the counts as input, producing derived information such as lists of successfully accessed pages per day.  This processing can be adjusted as needed, without the need to refer back to the original source log files (which can then be deleted if needed without affecting the future options for analysis).

### CODE REVIEW

File: lib/count_cdn_logs.rb [hunk 1]
```diff
@@ -0,0 +1,36 @@
+# coding: utf-8
```
Comment (tijmenb) [inline lib/count_cdn_logs.rb]: Ruby 2.0 is utf-8 by default, this isn't neccessary anymore if we add the appropriate `.ruby-version`

File: lib/count_cdn_logs.rb [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+
+class CountCdnLogs
+  def initialize(log_dir, processed_dir)
+    @log_dir = log_dir
+    @processed_dir = processed_dir
+    @counts_dir = "#{processed_dir}/raw_counts"
+  end
+
+  def update
+    files = completed_log_files
```
Comment (tijmenb) [inline lib/count_cdn_logs.rb]: This line is redundant if you change `files.each` to `completed_log_files.each`

File: lib/first_last_success_calculator.rb [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+
+  def process
+    remove_out_of_date_output
+    update_first_last
+  end
+
+private
+
+  def current_daily_files_and_sizes
+    @current_daily_files_and_sizes ||=
```
Comment (tijmenb) [inline lib/first_last_success_calculator.rb]: Matybe having these on the same line could be confusing, perhaps wrapping the code in `begin` & `end` would help.

File: lib/first_last_success_calculator.rb [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+        path, stored_size = line.strip.split(/ /, 2)
+        result[path] = stored_size.to_i
+      }
+    else
+      {}
+    end
+  end
+
+  def remove_out_of_date_output
+    if source_files_have_changed_or_gone
```
Comment (tijmenb) [inline lib/first_last_success_calculator.rb]: Better with a question mark: `source_files_have_changed_or_gone?`

File: lib/known_good_calculator.rb [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+# to the list.
+
+require_relative 'config_logging'
+
+class KnownGoodCalculator
+  attr_reader :output_dir
+  attr_reader :first_last_file
+  attr_reader :known_good_urls_file
+
+  REQUIRED_SEPARATION_DAYS = 7
```
Comment (tijmenb) [inline lib/known_good_calculator.rb]: Is this the number of days a page should have existed?

File: lib/known_good_calculator.rb [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+  def initialize(processed_dir)
+    successes_dir = "#{processed_dir}/successes"
+    @output_dir = "#{processed_dir}/output"
+    @first_last_file = "#{successes_dir}/first_last"
+    @known_good_urls_file = "#{output_dir}/known_good_urls"
+  end
+
+  def process
+    ensure_directories_exist
+    unless already_up_to_date
```
Comment (tijmenb) [inline lib/known_good_calculator.rb]: `unless already_up_to_date?` reads better.

File: lib/count_cdn_logs.rb (hunk 2 above)
Comment (rboulton) [inline lib/count_cdn_logs.rb]: I was actually wanting to avoid evaluating completed_log_files twice, in case the result was changed (eg, by an accidentally concurrently running program). But I failed to actually do that.

Instead, I'm changing this to memoize the result of completed_log_files, as well as calling it directly as suggested.
